### PR TITLE
Upsert in Postgres is natively supported in 9.5 & 9.6

### PR DESCRIPTION
**Pull request by nremond**
Hi,

The two last versions of Postgresql natively support upsert query, cf: (https://www.postgresql.org/docs/9.5/static/sql-insert.html#SQL-ON-CONFLICT)

I haven't found the associated test and how to run tests on the Postgres database, could someone help me with that?

So, just starting the discussion ...

### CODE REVIEW

File: slick/src/main/scala/slick/jdbc/PostgresProfile.scala [hunk 1]
```diff
@@ -177,11 +177,11 @@ trait PostgresProfile extends JdbcProfile {
       val nonAutoIncNames = nonAutoIncSyms.map(fs => quoteIdentifier(fs.name)).mkString(",")
       val nonAutoIncVars = nonAutoIncSyms.map(_ => "?").mkString(",")
       val cond = pkNames.map(n => s"$n=?").mkString(" and ")
-      val insert = s"insert into $tableName ($nonAutoIncNames) select $nonAutoIncVars where not exists (select 1 from $tableName where $cond)"
-      new InsertBuilderResult(table, s"$update; $insert", ConstArray.from(softSyms ++ pkSyms))
+      val upsert = s"""insert into $tableName ($nonAutoIncNames) select $nonAutoIncVars on conflict (${softNames.mkString(",")}) do $update"""
```
Comment (cvogt) [inline slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: do we need the where condition still in `update` in this case?

**Comment by cvogt** [comment]
LGTM! @tminglei mind giving this a second review?

File: slick/src/main/scala/slick/jdbc/PostgresProfile.scala (hunk 1 above)
Comment (nremond) [inline slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: You're right, from the doc : 

```sql
INSERT INTO distributors (did, dname)
    VALUES (5, 'Gizmo Transglobal'), (6, 'Associated Computing, Inc')
    ON CONFLICT (did) DO UPDATE SET dname = EXCLUDED.dname;
```

Commit updated

**Comment by nremond** [comment]
@cvogt how can I execute `slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/InsertTest.scala` with Postgres ?

File: slick/src/main/scala/slick/jdbc/PostgresProfile.scala [hunk 2]
```diff
@@ -173,15 +173,12 @@ trait PostgresProfile extends JdbcProfile {
 
   class UpsertBuilder(ins: Insert) extends super.UpsertBuilder(ins) {
     override def buildInsert: InsertBuilderResult = {
-      val update = "update " + tableName + " set " + softNames.map(n => s"$n=?").mkString(",") + " where " + pkNames.map(n => s"$n=?").mkString(" and ")
-      val nonAutoIncNames = nonAutoIncSyms.map(fs => quoteIdentifier(fs.name)).mkString(",")
-      val nonAutoIncVars = nonAutoIncSyms.map(_ => "?").mkString(",")
-      val cond = pkNames.map(n => s"$n=?").mkString(" and ")
-      val insert = s"insert into $tableName ($nonAutoIncNames) select $nonAutoIncVars where not exists (select 1 from $tableName where $cond)"
-      new InsertBuilderResult(table, s"$update; $insert", ConstArray.from(softSyms ++ pkSyms))
+      val update = s"""update $tableName set ${softNames.map(n => s"$n=EXCLUDED.$n").mkString(",")}"""
```
Comment (tminglei) [inline slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: `$tableName` should be removed

File: slick/src/main/scala/slick/jdbc/PostgresProfile.scala [hunk 3]
```diff
@@ -173,15 +173,12 @@ trait PostgresProfile extends JdbcProfile {
...
+      val upsert = s"""insert into $tableName (${allNames.mkString(",")}) select $allVars on conflict (${pkNames.mkString(",")}) do $update"""
+      new InsertBuilderResult(table, s"$upsert", ConstArray.from(softSyms))
```
Comment (tminglei) [inline slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: `ConstArray.from(softSyms)` should be `syms`

File: slick/src/main/scala/slick/jdbc/PostgresProfile.scala [hunk 4]
```diff
@@ -173,15 +173,12 @@ trait PostgresProfile extends JdbcProfile {
...
     }
 
-    override def transformMapping(n: Node) = reorderColumns(n, softSyms ++ pkSyms ++ nonAutoIncSyms.toSeq ++ pkSyms)
+    override def transformMapping(n: Node) = reorderColumns(n, syms.toSeq ++ softSyms)
```
Comment (tminglei) [inline slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: `syms.toSeq ++ softSyms` should be `syms.toSeq`

Comment (nremond) [inline slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: Fixed.

File: slick/src/main/scala/slick/jdbc/PostgresProfile.scala (hunk 3 above)
Comment (nremond) [inline slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: Fixed.

File: slick/src/main/scala/slick/jdbc/PostgresProfile.scala (hunk 2 above)
Comment (nremond) [inline slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: Fixed.

**Comment by nremond** [comment]
@tminglei Thanks for the review. I didn't see `slick-pg` already had that code. I feel like considering autoInc column that isn't the primary key like a fair concern and we should probably just import your code, no?

**Comment by tminglei** [comment]
@nremond you can refer to it, but your implementation is more clear.

**Comment by cvogt** [comment]
Just realized this is only for the latest two versions. What's the advantage of using the native syntax rather than Slick's encoding (which works across more versions) @nremond ?

**Comment by nremond** [comment]
> "UPSERT" is a DBMS feature that allows a DML statement's author to atomically either insert a row, or on the basis of the row already existing, UPDATE that existing row instead, while safely giving little to no further thought to concurrency.
https://wiki.postgresql.org/wiki/UPSERT#.22UPSERT.22_definition

So, @cvogt, the difference is that Slick's version is *not atomic*. I would personally not even call it upsert as it's very misleading.

**Comment by cvogt** [comment]
@nremond ah, right. Good point. Can we fall back to the non-atomic one on postgres < 9.5, just so we do not break compatibility?

**Comment by nremond** [comment]
@cvogt For sure. How would you do the fallback mechanism? How to detect which version of postgres it is?

**Comment by tminglei** [comment]
@nremond slick-pg allows user to enable native upsert by adding back `JdbcProfile.capabilities.insertOrUpdate`.

For its usage and implementation, you can refer to its [tests](https://github.com/tminglei/slick-pg/blob/master/core/src/test/scala/com/github/tminglei/slickpg/PgUpsertSuite.scala#L15) and [codes](https://github.com/tminglei/slick-pg/blob/master/core/src/main/scala/com/github/tminglei/slickpg/ExPostgresDriver.scala#L20).

**Comment by nremond** [comment]
For sure. How would you do the fallback mechanism? How to detect which
version of postgres it is?

**Comment by tminglei** [comment]
@nremond I didn't determine postgres version, instead I let user specify it by by [adding back `JdbcProfile.capabilities.insertOrUpdate`](https://github.com/tminglei/slick-pg/blob/master/core/src/test/scala/com/github/tminglei/slickpg/PgUpsertSuite.scala#L15), if he was sure the native upsert is supported.

**Comment by cvogt** [comment]
@nremond should we go with @tminglei's strategy? Otherwise we'd have to connect to the server to find out the version, which is not ideal. Instead of letting the user override the field by hand, we could have a second driver object built-in:
```
  object Postgres_9_5_OrLater extends PostgresDriver {
    // Add back `capabilities.insertOrUpdate` to enable native `upsert` support
    override protected def computeCapabilities: Set[Capability] =
      super.computeCapabilities + JdbcProfile.capabilities.insertOrUpdate
  }
```
or something like that and document it as an alternative for 9.5 or later.

**Comment by nremond** [comment]
@tminglei @cvogt Thank you both for your inputs. I agree that doing something at runtime is not ideal. I'll work on integrating that ASAP.

**Comment by szeiger** [comment]
I suggest making it a configuration option of the driver. We already do that in other cases for default text types. Don't forget to make it work without reference.conf, otherwise it will fail in macro-based code.

**Comment by lforite** [comment]
Hello, any plan on merging this PR ?

**Comment by cvogt** [comment]
the travis build fails

File: slick/src/main/scala/slick/jdbc/PostgresProfile.scala [hunk 5]
```diff
@@ -173,15 +173,13 @@ trait PostgresProfile extends JdbcProfile {
 
   class UpsertBuilder(ins: Insert) extends super.UpsertBuilder(ins) {
     override def buildInsert: InsertBuilderResult = {
-      val update = "update " + tableName + " set " + softNames.map(n => s"$n=?").mkString(",") + " where " + pkNames.map(n => s"$n=?").mkString(" and ")
-      val nonAutoIncNames = nonAutoIncSyms.map(fs => quoteIdentifier(fs.name)).mkString(",")
-      val nonAutoIncVars = nonAutoIncSyms.map(_ => "?").mkString(",")
-      val cond = pkNames.map(n => s"$n=?").mkString(" and ")
-      val insert = s"insert into $tableName ($nonAutoIncNames) select $nonAutoIncVars where not exists (select 1 from $tableName where $cond)"
-      new InsertBuilderResult(table, s"$update; $insert", ConstArray.from(softSyms ++ pkSyms))
+      val insert = s"""insert into $tableName (${allNames.mkString(",")}) values $allVars"""
```
Comment (lforite) [inline slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: Isn't MERGE command more idiomatic, as part of SQL standard ? See https://wiki.postgresql.org/wiki/UPSERT

Comment (nremond) [reply slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: https://www.postgresql.org/docs/9.6/static/sql-commands.html I don't see any `MERGE` syntax? I think Postgres only support `INSERT ON CONFLICT` or am I missing something?

**Comment by nremond** [comment]
@cvogt The circleCI build passed. Travis failed with the following:
```
The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends --force-yes install graphviz python-sphinx mysql-server" failed and exited with 100 during .
```

No clue on what I can do for that :-(

File: slick/src/main/scala/slick/jdbc/PostgresProfile.scala (hunk 5 above)
Comment (lforite) [reply slick/src/main/scala/slick/jdbc/PostgresProfile.scala]: yep my bad, I got confused by the wiki !

**Comment by hvesalai** [comment]
Please rebase this to master to get the tests to run

**Comment by SethTisue** [comment]
closing for inactivity

**Pull request closed without merging**